torch/csrc/utils/tensor_new.cpp

```diff
@@ -236,7 +236,8 @@ Tensor internal_new_from_data(
     bool copy_variables,
     bool copy_numpy,
     bool type_inference,
-    bool pin_memory = false) {
+    bool pin_memory = false,
+    c10::DeviceIndex device_index = -1) {
```

Comment:
having both `device_index` and `device_opt` arguments with unclear precedence, and computing actual device based not only on these 2 arguments, but also on `type_inference`, is super confusing.

Reply:
Thanks for the feedback @ngimel.   

> So behavior introduced by this PR (unless I'm misreading) is:
> 
> 1. if device is specified, the sparse tensor will be created on this device.
Yes, It will.  Nothing changed here.

> 2. if device is unspecified, and dtype is also unspecified, the sparse tensor will be created with dtype and device of vals.
Yes, It will. Basically, this fix cover this scenario by using the same device information of the vals tensor including the `device_index` for the creation of the indices tensor. Before this change, the function `internal_new_from_data` assumed that the vals tensor was by default at 'cuda:0' and it failed when values tensor was set up at 'cuda:1' for example.  

> 3. if device is unspecified, but dtype is specified, the sparse tensor will be created on default device, and device of val will be ignored. Why difference with the previous case? Is it documented somewhere?
Yes the device info will be ignored and will use by default 'cpu' in this case the `device_index` is optional and unnecessary.   

>> having both device_index and device_opt arguments with unclear precedence, and computing actual device based not only on these 2 arguments, but also on type_inference, is super confusing.

I agree with that. Let me refactor this just a little. The Parameter should be a tuple of  <c10::DispatchKey, c10::DeviceIndex> to identify the default `dispatch_key` device. So, in case that vals tensor is set up for example at device 'cuda:1', the indices tensor should be in the same device.

Reply:
Let's step back and spell out expected behavior though. I propose if device or dtype is unspecified, fill it in from vals, which is not what this PR does. So
1) device and dtype are specified - use them
2) device and dtype are unspecified - fill in from vals
3) device is specified, dtype is unspecified - use device, fill in dtype from vals
4) device is unspecified, dtype is specified - fill in device from vals, use dtype
Does it make sense?

Reply:
This seems OK, but for safety what if the function throws an error when the device is unspecified and the indices and value tensors have a device mismatch? So a slight change to:

1) if device specified, use it; if dtype is specified, use it
2) if device is unspecified, check that both indices and values are on the same device, if they are then use that device, if they're not then throw a Runtime error ("Expected indices and values to be on the same device, but got ..."
3) if dtype is unspecified, use the dtype of the values tensor

Reply:
> Let's step back and spell out expected behavior though. I propose if device or dtype is unspecified, fill it in from vals, which is not what this PR does. So
> 
> 1. device and dtype are specified - use them
All right. No changes required for this case.

> 2. device and dtype are unspecified - fill in from vals
All right. This fix cover this case. 

> 3. device is specified, dtype is unspecified - use device, fill in dtype from vals
All right. No changes required for this case.

> 4. device is unspecified, dtype is specified - fill in device from vals, use dtype
All right. No changes required for this case.

>    Does it make sense?

Reply:
> This seems OK, but for safety what if the function throws an error when the device is unspecified and the indices and value tensors have a device mismatch? So a slight change to:

That is a good question. Inside the function `sparse_coo_tensor_ctor`  function `internal_new_from_data` is called twice once for values and then one for indices tensors. If there is a mismatch the function `internal_new_from_data`  will assure that both (values and indices) are going to be in the same device by calling `return var.to(device, inferred_scalar_type, ...)`/ 

> 
> 1. if device specified, use it; if dtype is specified, use it

All right. No changes required for this case.

> 2. if device is unspecified, check that both indices and values are on the same device, if they are then use that device, if they're not then throw a Runtime error ("Expected indices and values to be on the same device, but got ..."

I have my concerns here. If I add this exception.  The code related to make sure that both values and indices are using the same device though the function  `return var.to(device, inferred_scalar_type, ...)` will be unnecessary, and we will lose this functionality. Moreover, some unit tests rely on this functionality.  

```
https://pytorch.org/docs/master/generated/torch.sparse_coo_tensor.html
device (torch.device, optional) – the desired device of returned tensor. Default: if None, uses the current device for the default tensor type (see torch.set_default_tensor_type()). device will be the CPU for CPU tensor types and the current CUDA device for CUDA tensor types.
```
So, maybe it is better to clarify in the TORCH.SPARSE_COO_TENSOR documentation the  expected behavior for this case. 

> 3. if dtype is unspecified, use the dtype of the values tensor
All right. No changes required for this case.

Looking forward your comments, cc @mruberry

Reply:
Re: the 4th case, my proposal differs from what's currently implemented, that's true. But IMO it makes it more consistent, in that missing data is always filled from the vals tensor. Otherwise, missing data is filled from the vals everywhere (cases 2 and 3) except when only device is missing, and then default device is used.

Reply:
Thanks for the feedback @ngimel. So just to be clear.  If device is unspecified, dtype is specified - fill in device from vals, use dtype. However, if device is unspecified and indices and values are not on the same device,  I suggest just printing a warning ("Expected indices and values to be on the same device, but got ..."   instead of a runtime error in order to maintain the same expected behavior as `torch.sparse.FloatTensor` constructor.  Also,  we will use the vals' device as a default device for indices. 

Looking forward your comments.

Reply:
Yes, this seems fine, warn if vals and indices are on the different devices, and create vals and indices on the vals device.